lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp

```diff
@@ -718,8 +718,24 @@ Operation *LayoutPropagation::rewriteOp(Operation *op) {
   return nullptr;
 }
 
-bool canBeRemat(Operation *op) {
-  if (isa<LoadOp, StoreOp>(op))
+bool WillBeMoreExpensiveLoad(LoadOp op, Attribute newEncoding) {
+  auto tensorType = op.getPtr().getType().cast<RankedTensorType>();
+  auto oldEncoding = tensorType.getEncoding();
+  auto oldContigPerThread =
+      triton::gpu::getUniqueContigPerThread(oldEncoding, tensorType.getShape());
+  auto oldOrder = triton::gpu::getOrder(oldEncoding);
+  auto oldContiguity = oldContigPerThread[oldOrder[0]];
+
+  // If the original load cannot be vectorized, the new load will likely not be
+  // more expensive.
+  return oldContiguity > 1;
```

Comment:
I think a similar example than what I mentioned before would show the problem. If we have a transpose kernel the load and the store may not be vectorized, yet we still want coalescing to happen to have efficient memory access patterns, this would cause the load to potentially read horizontally which would tank the perf.

Reply:
Do you mean that we could undo the decision made by the coalescing pass? What would that generated layout look like, if not vectorized?

BTW, there was a transpose kernel example in test_core.py, i.,e, `test_trans_reshape`, which was failing and not works fine with the latest version.

Reply:
> Do you mean that we could undo the decision made by the coalescing pass? What would that generated layout look like, if not vectorized?

right, 

For instance you may have IR like this:
```
#blocked = #triton_gpu.blocked<{sizePerThread = [1, 1], threadsPerWarp = [1, 32], warpsPerCTA = [1, 1], order = [1, 0]}>
#blocked = #triton_gpu.blocked<{sizePerThread = [1, 1], threadsPerWarp = [32, 1], warpsPerCTA = [1, 1], order = [0, 1]}>
...
%l = tt.load %p {cache = 1 : i32, evict = 1 : i32, isVolatile = false} : tensor<32x32xf32, #blocked>
%c = triton_gpu.convert_layout %l : tensor<32x32xf32> -> tensor<32x32xf32, #blocked1> loc(#loc60)
tt.store %tp, %c {cache = 1 : i32, evict = 1 : i32} : tensor<32x32xf32, #blocked1>
```
Here we are reading along the inner dimension as we expect the `%p` to be contiguous along this dimension and we are writing going along the outer dimension as we expect `%tp` to be contiguous along this dimension. 
If you write a simple transpose kernel (without using tl.trans, just loading along 1 dimension and storing along a different one) you should be able to see the perf.

> BTW, there was a transpose kernel example in test_core.py, i.,e, test_trans_reshape, which was failing and not works fine with the latest version.

No this is unrelated.

Reply:
I see, thanks for the example!

It does sound very tricky to come up with a reliable heuristics, but in this case can we check in addition that the original layout has the same order with the new layout so that coalescing will work similarly for both? In my case, it'd be checking if the new layout is a mma layout and the original layout is a row-major.

Reply:
coalescing won't be the same for mma layout and blocked layout right? I think we need some kind of cost model to figure out that skipping the convert is better but I'm not sure what it would look like.

If you can share a bit more about your example maybe we can bounce ideas on something that make sense. I think in this case even if check that the order match it is not clear that the transformation is beneficial.

Reply:
Yeah, mma layout isn't the same with blocked layout in terms of coalescing.

I added my example as a test case in this PR. It is basically a biased attention kernel. I'm trying to convert the three loads with `blocked1` layout to `mma` layout.

```
#blocked1 = #triton_gpu.blocked<{sizePerThread = [1, 1], threadsPerWarp = [16, 2], warpsPerCTA = [1, 4], order = [0, 1]}>
#mma = #triton_gpu.nvidia_mma<{versionMajor = 2, versionMinor = 0, warpsPerCTA = [1, 4], instrShape = [16, 8]}>
```